src/Atmos/Model/source.jl

```diff
@@ -173,3 +181,69 @@ function atmos_source!(
         source.ρu -= β_sponge * (state.ρu .- state.ρ * s.u_relaxation)
     end
 end
+
+"""
+    RemovePrecipitation{FT} <: Source
+A sink to `q_tot` when cloud condensate is exceeding a threshold.
+The threshold is defined either in terms of condensate or supersaturation.
+The removal rate is implemented as a relaxation term
+in the Microphysics_0M module.
+The default thresholds and timescale are defined in CLIMAParameters.jl.
+"""
+struct RemovePrecipitation <: Source
+    " Set to true if using qc based threshold"
+    use_qc_thr::Bool
+end
+function atmos_source!(
+    s::RemovePrecipitation,
+    atmos::AtmosModel,
+    source::Vars,
+    state::Vars,
+    diffusive::Vars,
+    aux::Vars,
+    t::Real,
+    direction,
+)
+    # TODO - how to assert that it is used with EquilMoist module?
+    #        (if it were to be used with non-EquilMoist model
+    #         q_liq and q_ice would be part of state)
+    FT = eltype(state)
+    if aux.moisture.q_liq + aux.moisture.q_ice > eps(FT)
+
+        q = PhasePartition(
+            state.moisture.ρq_tot / state.ρ,
+            aux.moisture.q_liq,
+            aux.moisture.q_ice,
+        )
+        T::FT = aux.moisture.temperature
+        lf::FT = liquid_fraction(atmos.param_set, T, PhaseEquil, q)
+
+        _e_int_i0::FT = e_int_i0(atmos.param_set)
+        _cv_d::FT = cv_d(atmos.param_set)
+        _cv_l::FT = cv_l(atmos.param_set)
+        _cv_i::FT = cv_i(atmos.param_set)
+        _T_0::FT = T_0(atmos.param_set)
+
+        S_qt::FT = 0
+        if s.use_qc_thr
+            S_qt = remove_precipitation(atmos.param_set, q)
+        else
+            ts_neq = TemperatureSHumNonEquil(atmos.param_set, T, state.ρ, q)
```

Comment:
Why do we need `TemperatureSHumNonEquil` instead of `TemperatureSHumEquil` here?

Reply:
I assume a non-equilibrium thermo state is needed. I think I'll add a convenience method for computing non-equilibrium thermo states since microphysics seems to need them even when using an equilibrium moist model.

Reply:
I'm a bit confused because we have `lf::FT = liquid_fraction(atmos.param_set, T, PhaseEquil, q)` here. I could be wrong though, it seems I need to familiarize myself again with Thermodynamics.jl!

Reply:
What I want to do is to compute saturation vapour pressure. In order to do that, I have to create the thermo state and pass it into `q_vap_saturation`. I'm not a big fan of that, but that is the general strategy when using functions from `Thermodynamics.jl` module. However, I think that if I pass into  `TemperatureSHumNonEquil` T, rho and q it will not do anything to them, just populate the thermo_state fields. This is what I want, because I just want to pass them as arguments to `q_vap_saturation`.

Reply:
For this, I think you could simply do:

```
ts = thermo_state(atmos, atmos.moisture, state, aux)
q_vap = q_vap_saturation(ts)
```
@charleskawczynski is this correct?

Reply:
What does it do? Will it create phase partition based on q_liq and q_ice saved in aux? Or will it do another saturation adjustment?

Reply:
I don't think it does SA, but it may return a `PhaseEquil` state. I guess I'm just confused why the liquid fraction is `PhaseEquil` but the thermodynamic state is `PhaseNonEquil`.

Reply:
I just treat them as containers for temperature and phase partition. Will try to understand what functions are being called when you do what you are suggesting. I want to make sure I don't call another saturation adjustment here.

Reply:
What @trontrytel has done is fine for now, but we should probably discuss if there is a better/more proper alternative to creating non-equilibrium states when we’re using an equilibrium model. `thermo_state` recovers (but does not itself call SA) the saved thermo state for whichever moisture model is being used.

Reply:
After re-reading this, I think we should actually call `thermo_state`. Let’s discuss this more @trontrytel

Reply:
If I understand correctly, `lf::FT = liquid_fraction(atmos.param_set, T, PhaseEquil, q)` will only return 1 or 0? But we would like to have phase partition based on q_liq and q_ice?

Reply:
> If I understand correctly, `lf::FT = liquid_fraction(atmos.param_set, T, PhaseEquil, q)` will only return 1 or 0? But we would like to have phase partition based on q_liq and q_ice?

It returns `q.liq / q_c` for `PhaseNonEquil` _if_ there is condensate. Otherwise, it returns `FT(T > T_freeze(param_set))`. So, in this case it returns `FT(T > T_freeze(param_set))`.

Reply:
IIUC, this thermo state will be exactly the same as an equilibrium thermo state.